### DISCUSSION TITLE
Dependency with main actor-isolated initializer requirements like FeedbackGeneratorClient from isowords

### DISCUSSION THREAD

**Discussion by ph1ps** [pointfreeco/swift-dependencies#22] [Q&A] [answered]
Hello,

I was trying to extract `UISelectionFeedbackGenerator` into its own dependency like it is in [isowords](https://github.com/pointfreeco/isowords/tree/main/Sources/FeedbackGeneratorClient).
Very similar to isowords, the client itself looks like this:

```swift
public struct FeedbackGeneratorClient: Sendable {
  
  public internal(set) var prepare: @Sendable () async -> Void
  public internal(set) var generate: @Sendable () async -> Void
  
  public init(
    prepare: @escaping @Sendable () async -> Void,
    generate: @escaping @Sendable () async -> Void
  ) {
    self.prepare = prepare
    self.generate = generate
  }
}

extension DependencyValues {
  
  public var feedbackGeneratorClient: FeedbackGeneratorClient {
    get { self[FeedbackGeneratorClient.self] }
    set { self[FeedbackGeneratorClient.self] = newValue }
  }
}
```

And the live implementation looks like this:

```swift
extension FeedbackGeneratorClient: DependencyKey {
  
  public static let liveValue = {
    
    let selectionFeedbackGenerator = UISelectionFeedbackGenerator() // 🔴 Call to main actor-isolated initializer 'init()' in a synchronous nonisolated context

    return Self(
      prepare: { await selectionFeedbackGenerator.prepare() },
      generate: { await selectionFeedbackGenerator.selectionChanged() }
    )
  }()
}
```

However if you have `-warn-concurrency` enabled, there will be a compilation error at the line I marked. I think I understand what the compiler is trying to tell me, however I did not find a nice solution yet on how to handle this (without disabling the concurrency "warnings" of course 😅). Has someone else experienced something similar to this yet?

Maybe something like a `MainActorIsolated` similar to `ActorIsolated` and `LockIsolated` would make sense here?

**Comment by tgrapperon** [comment, marked as answer]
Hey @ph1ps, you're right, it is indeed really annoying to wrap main actor stuff with the protocol requirement getting in the way.
I have a `MainActorIsolated` wrapper that I'm toying with to tackle these issues:
```swift
public final class MainActorIsolated<Value: AnyObject>: @unchecked Sendable {
  @MainActor
  public private(set) lazy var value: Value = initialValue()
  private let initialValue: @MainActor @Sendable () -> Value
  public init(_ initialValue: @MainActor @Sendable @escaping () -> Value) {
    self.initialValue = initialValue
  }
  
  @MainActor
  public func withValue<T: Sendable>(
    _ operation: @MainActor (inout Value) throws -> T
  ) rethrows -> T {
    return try operation(&value)
  }
  
  @MainActor
  public subscript<Subject: Sendable>(dynamicMember keyPath: KeyPath<Value, Subject>) -> Subject {
    self.value[keyPath: keyPath]
  }
}
```
Because the value is only lazily created in a `MainActor` context, you can create values of this type synchronously everywhere. It would allow to conveniently wrap `UISelectionFeedbackGenerator()` as `MainActorIsolated({ UISelectionFeedbackGenerator() })` (I think that we can't use an `@autoclosure`, but I may be wrong).

I was planning to add this utility to [`swift-dependencies-additions`](https://github.com/tgrapperon/swift-dependencies-additions) very soon, but if Brandon and Stephen prefer, I can directly PR on `swift-dependencies`.

**Reply by ph1ps** [reply]
Ah yes, that looks way nicer than what I had. May I ask why `Value` is constrained to being `AnyObject`? Not that I need otherwise, but just curious.

**Reply by tgrapperon** [reply]
Because it represents the case I had at hand. But we can probably refactor to accept `MainActor` value types using the same style as `LockIsolateld`.
As I'm thinking more about it, we can probably expose the lazy property safely as a read-write one, and get rid of the `withValue`.

**Reply by tgrapperon** [reply]
This very minimal version looks like to be enough to build witness-style dependencies:
```swift
public final class MainActorIsolated<Value: Sendable>: @unchecked Sendable {
  @MainActor
  public lazy var value: Value = initialValue()
  private let initialValue: @MainActor @Sendable () -> Value
  public init(initialValue: @MainActor @Sendable @escaping () -> Value) {
    self.initialValue = initialValue
  }
}
```

**Reply by ph1ps** [reply]
I am wondering why there is a warning when you remove the `@unchecked`:
> 🟡 Stored property '$__lazy_storage_$_value' of 'Sendable'-conforming generic class 'MainActorIsolated1' is mutable

Actor isolated vars are safe in this context, aren't they? Maybe a frontend bug for lazy vars?

**Reply by tgrapperon** [reply]
Yeah, I'm not really sure if the `lazy` assignation will happen on the `MainActor` or not, so the `unchecked` is maybe unjustified. But this can be refactored using a computed property:
```swift
public final class MainActorIsolated<Value: Sendable>: @unchecked Sendable {
  @MainActor // I don't think we need this.
  private var _value: Value!
  @MainActor
  public var value: Value {
    get {
      if _value == nil { _value = initialValue() }
      return _value
    }
    set { _value = newValue }
  }
  private let initialValue: @MainActor @Sendable () -> Value
  public init(initialValue: @MainActor @Sendable @escaping () -> Value) {
    self.initialValue = initialValue
  }
}
```
We still need the `@unchecked` annotation, but this one is safe I guess.

**Reply by tgrapperon** [reply]
Alternatively, this one is `Sendable` without any warning:
```swift
@MainActor
public final class MainActorIsolated<Value>: Sendable {
  public lazy var value: Value = initialValue()
  private let initialValue: @MainActor () -> Value
  nonisolated public init(initialValue: @MainActor @escaping () -> Value) {
    self.initialValue = initialValue
  }
}
```
https://github.com/apple/swift-evolution/blob/main/proposals/0316-global-actors.md#using-global-actors-on-a-type
> A non-protocol type that is annotated with a global actor implicitly conforms to Sendable. Instances of such types are safe to share across concurrency domains because access to their state is guarded by the global actor.

**Reply by filiplazov** [reply]
Hey @tgrapperon have you used this in your projects so far?
I am hit with a similar use case (not completely same) where I want to have a shared state dependency like 

```swift
@MainActor
final class SharedStuff {
  var someInt: Int
}
```

The main purpose of this dependency is to share some state across many reducers but I want to restrict mutation to the main actor. What is the recommended approach for this?

**Comment by kamcma** [comment]
> Alternatively, this one is `Sendable` without any warning:
> 
> ```swift
> @MainActor
> public final class MainActorIsolated<Value>: Sendable {
>   public lazy var value: Value = initialValue()
>   private let initialValue: @MainActor () -> Value
>   nonisolated public init(initialValue: @MainActor @escaping () -> Value) {
>     self.initialValue = initialValue
>   }
> }
> ```

Hello all. Been working on a dependency with the same issue: needing to initialize resources on the main actor. Unfortunately, the quoted suggestion is now warning for me on Xcode 15.3 / Swift 5.10 / `.enableExperimentalFeature("StrictConcurrency")`:

```swift
@MainActor
public final class MainActorIsolated<Value>: Sendable {
  public lazy var value: Value = initialValue() // ⚠️ Non-sendable type '@MainActor () -> Value' in asynchronous access to main actor-isolated property 'initialValue' cannot cross actor boundary
  private let initialValue: @MainActor () -> Value
  nonisolated public init(initialValue: @MainActor @escaping () -> Value) {
    self.initialValue = initialValue // ⚠️ Main actor-isolated property 'initialValue' can not be mutated from a non-isolated context; this is an error in Swift 6
  }
}
```

Was playing around, trying to come up with a solution to this problem. What about something like this? Gives me no warnings.

```swift
struct LocationClient: Sendable {
  var requestLocation: @Sendable () async -> Void
  // etc
}

extension LocationClient: DependencyKey {
  public static var liveValue: Self {
    Self(
      requestLocation: { @MainActor in
        Self.manager.requestLocation()
      }
      // etc
    )
  }

  @MainActor
  private static let manager = {
    let manager = CLLocationManager()
    // configure resources if necessary
    return manager
  }()
}
```
Not sure if there's a way to extract the main-actor-static isolation into a wrapper type that doesn't warn; could save needing to annotate every endpoint closure `@MainActor`

Got the idea to store resources as statics on the client from [isowords' Game Center client](https://github.com/pointfreeco/isowords/blob/ca796b78ef4d9af4bac0e7db2b1c08b48d82e7a5/Sources/ComposableGameCenter/LiveKey.swift#L220-L221).

**Reply by ph1ps** [reply]
To fix the warnings you could just mark the closures @Sendable like this:

```swift
@MainActor
public final class MainActorIsolated<Value>: Sendable {
  public lazy var value: Value = initialValue()
  private let initialValue: @Sendable @MainActor () -> Value
  nonisolated public init(initialValue: @escaping @Sendable @MainActor () -> Value) {
    self.initialValue = initialValue
  }
}
```

But I like your approach. Definitely simpler.

**Reply by kamcma** [reply]
Ah, thank you for the pointer about fixing those warnings. Alright, well now I have some options. I will continue to think on it.

**Reply by kamcma** [reply]
Important caveat on my solution for anyone reading along at home: apparently `static let` main actor isolation is broken before Swift 5.10: apple/swift#69607

**Comment by jtbergman** [comment]
Adding on to an old discussion, but this solution seems to work in Swift 6 language mode. I'm using this solution to share a TCA store via `@Dependency` which also faces the `MainActor` isolated initializer issue. 

I introduced this type which is not isolated:
```swift
struct Provider<T>: Sendable {
  private let provide: @MainActor () -> T
  
  init(_ provide: @autoclosure @MainActor @escaping () -> T) {
    self.provide = provide
  }
  
  @MainActor func callAsFunction() -> T {
    provide()
  }
}
```

Then I register the dependency like any other.
```swift
extension DependencyValues {
  var feature: Provider<StoreOf<Feature>> {
    get { self[Feature.self] }
    set { self[Feature.self] = newValue }
  }
}

extension Feature: DependencyKey {
  // Need to use a static instance for the live value
  @MainActor private static let shared = StoreOf<Feature>(initialState: Feature.State()) { 
    Feature()
  }

  static let liveValue: Provider<StoreOf<Feature>> = Provider(Self.shared)
}
```

And finally, it can be used anywhere, but the key path must be invoked (i.e. `\.feature()`):
```swift
// Usage off MainActor
final class MainActorDependency {
  @Dependency(\.feature()) var feature
  
  func testing() async {
    print(await feature.state)
    await feature(.action)
  }
}

// Usage on MainActor
@MainActor 
final class SimpleUseFromMain {
  @Dependency(\.feature()) var feature
  
  func executesOnMain() {
    print(feature.state)
    feature.send(.action)
  }
}
```

**Reply by mbrandonw** [reply]
Hi @jtbergman, have you found that this actually works at runtime? It just crashes for me instantly:

<img width="806" height="141" alt="image" src="https://github.com/user-attachments/assets/22c65156-949a-4d6f-be49-215555a1edd3" />

Most likely just a Swift bug, but are you not seeing the same crash?

**Reply by jtbergman** [reply]
Oof, somewhere in my testing I switched to recompiling the module to save time and missed the runtime issue that was introduced 😅 I was actually surprised the code compiled after adding the `callAsFunction` and doing `\.feature()`, so I guess it makes sense it's causing a runtime crash. I'm guessing it's a Swift bug that it is allowed to compile. 

The approach can work at runtime, but it's no longer convenient. It just requires invoking `feature()` on a separate line.
```swift
// MainActor - won't crash at runtime
Task { @MainActor in
  @Dependency(\.feature) var feature
  let store = feature()
  store.send(.greet)
  print(await store.state)
}

// Detached – won't crash at runtime
Task.detached {
  @Dependency(\.feature) var feature
  let store = await feature()
  await store.send(.greet)
  print(await store.state)
}
```

Access to properties can be made more convenient using `@dynamicMemberLookup`.
```swift
@dynamicMemberLookup
struct Provider<T: Sendable>: Sendable {
  ...

  @MainActor var value: T {
    provide()
  }
  
  @MainActor subscript<U>(dynamicMember keyPath: KeyPath<T, U>) -> U {
    value[keyPath: keyPath]
  }
}

// Only makes MainActor access more convenient
Task { @MainActor in 
  @Dependency(\.feature) var feature
  print(feature.state) // No need to invoke feature()
}

// Off MainActor
// ❌ Non-Sendable type 'KeyPath<Store<Feature.State, Feature.Action>, Feature.State>' cannot be sent into main actor-isolated context in call to subscript 'subscript(dynamicMember:)'
Task.detached { 
  @Dependency(\.feature) var feature
  print(await feature.state)
}
```

And you can't call methods conveniently without accessing `value` directly or invoking `feature()` to get the provided dependency. In the case of TCA stores, it would be nice to provide something like the following, but I think that relies on [parameterized extensions](https://forums.swift.org/t/parameterized-extensions/25563), so it's not possible today.
```swift
// Not possible today? 
extension<S, A> Provider where T = Store<S, A>  {
  @MainActor func send(_ action: A)
    value.send(action)
  }
}
```

So I guess if I had to use this approach, I would use the original code without the dynamic member lookup, and just grabbing the dependency on a separate line so it's consistent on/off main and accessing properties/invoking methods. But definitely not as convenient anymore :/

Thanks for flagging the issue 🙏

**Reply by mbrandonw** [reply]
> ```swift
>   let store = feature()
>   store.send(.greet)
> ```

You have to be careful with this syntax because it could lead you to creating a whole new store each time `feature()` is invoked. You will want to make sure your `Provider` only captures a fully realized value and not a lazily created value to avoid that, and that may defeat the purpose.



> ```swift
> // Not possible today? 
> extension<S, A> Provider where T = Store<S, A>  {
>   @MainActor func send(_ action: A)
>     value.send(action)
>   }
> }
> ```

You can write this as:

```swift
extension Provider {
  @MainActor func send<S, A>(_ action: A) where T == Store<S, A> {
    value.send(action)
  }
}
```

But I'm still unsure of the overall pattern being employed here.
